pkg/controller/statefulset/stateful_set_control.go

```diff
@@ -574,35 +576,96 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		return &status, nil
 	}
 
+	return updateStatefulSetAfterInvariantEstablished(ctx,
+		ssc,
+		set,
+		replicas,
+		updateRevision,
+		status,
+	)
+
+}
+
+func updateStatefulSetAfterInvariantEstablished(
+	ctx context.Context,
+	ssc *defaultStatefulSetControl,
+	set *apps.StatefulSet,
+	replicas []*v1.Pod,
+	updateRevision *apps.ControllerRevision,
+	status apps.StatefulSetStatus,
+) (*apps.StatefulSetStatus, error) {
+
+	replicaCount := int(*set.Spec.Replicas)
+
 	// we compute the minimum ordinal of the target sequence for a destructive update based on the strategy.
 	updateMin := 0
+	maxUnavailable := 1
+	var err error
 	if set.Spec.UpdateStrategy.RollingUpdate != nil {
 		updateMin = int(*set.Spec.UpdateStrategy.RollingUpdate.Partition)
+
+		// if the feature was enabled and then later disabled, MaxUnavailable may have a value
+		// other than 1. Ignore the passed in value and Use maxUnavailable as 1 to enforce
+		// expected behavior when feature gate is not enabled.
+		if utilfeature.DefaultFeatureGate.Enabled(features.MaxUnavailableStatefulSet) {
+			maxUnavailable, err = intstrutil.GetValueFromIntOrPercent(intstrutil.ValueOrDefault(set.Spec.UpdateStrategy.RollingUpdate.MaxUnavailable, intstrutil.FromInt(1)), int(replicaCount), false)
+			if err != nil {
+				return &status, err
+			}
+
+		}
 	}
-	// we terminate the Pod with the largest ordinal that does not match the update revision.
+
+	// Collect all targets in the range between the partition and Spec.Replicas. Count any targets in that range that are unhealthy
+	// i.e. terminated or not running and ready as unavailable). Select the (MaxUnavailable - Unavailable) Pods, in order with respect
+	// to their ordinal for termination. Delete those pods and count the successful deletions. Update the status with the correct
+	// number of deletions.
+	unavailablePods := 0
 	for target := len(replicas) - 1; target >= updateMin; target-- {
+		if !isHealthy(replicas[target]) {
+			unavailablePods++
+		}
+	}
+
+	// Now we need to delete MaxUnavailable- unavailablePods
+	// start deleting one by one starting from the highest ordinal first
+	podsToDelete := maxUnavailable - unavailablePods
+	if podsToDelete < 0 {
+		podsToDelete = 0
+	}
+
+	deletedPods := 0
+	for target := len(replicas) - 1; target >= updateMin && podsToDelete > 0; target-- {
```

Comment:
If you follow my suggestion you don't need the `podsToDelete > 0 check`. But you could replace it with a `deletedPods < podsToDelete` check and lose an if statement inside the loop.

Reply:
i loved this suggestion, thanks.